src/core/thread/indirect_sender.cpp

```diff
@@ -125,6 +125,14 @@ Error IndirectSender::RemoveMessageFromSleepyChild(Message &aMessage, Child &aCh
 
     VerifyOrExit(aMessage.GetChildMask(childIndex), error = kErrorNotFound);
 
+    if (&aMessage == aChild.GetIndirectMessage())
+    {
+        aChild.ResetIndirectTxAttempts();
+#if OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE
+        aChild.ResetCslTxAttempts();
+#endif
+    }
```

Comment:
Thanks @edmont for finding and fixing this issue. I agree we should reset the tx attempt counter.

I would like to suggest a slightly different way to fix this such that we reset the attempt counters in `DataPollHandler` before we call `mCallbacks.HandleFrameChangeDone()`.
- I see we call the `HandleFrameChangeDone()` in 3 places and in two of them we do `ResetIndirectTxAttempts()` but not when it is directly called from `RequestFrameChange()`
- So maybe we add this for the case above and also maybe adding the `ResetCslTxAttempts()` for all cases? (maybe helpful if we add a new helper method that resets both attempt counters and use it instead?)

The reason I am suggesting this is the way we have broken the logic into `IndirectSender` and `DataPollHandler`:
- The idea is at some point we may move `DataPollHandler` to run on RCP, while `IndirectSender` manages it from host side
- That's why the interaction between these two modules uses async "request" and "callback" model. We have `RequestFrameChange()` to `kReplaceFrame` or `kPurgeFrame` and then callback to indicate when it is done.
- The `IndirectTxAttempts` is managed by `DataPollHandler` class and so we don't want `IndirectSender` to be able to access or reset it directly

Reply:
Makes sense, thanks. I was not aware about that design strategy.

I've added two commits:

1. On the first one, just moved resetting attempts into `HandleFrameChangeDone()`. It has the same effect, but maybe less "clean".
2. On the second, added a helper method as you suggested. But code size increases.

Let me know what's preferred.